src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs

```diff
@@ -506,6 +506,7 @@ namespace Microsoft.AspNetCore.Http.Generated
     {{GeneratedCodeAttribute}}
     file static class GeneratedRouteBuilderExtensionsCore
     {
+        private static readonly JsonOptions FallbackJsonOptions = new();
```

Comment:
You should try setting a resolver here, otherwise the singleton itself would be susceptible to races as multiple threads attempt to initialize it.

Reply:
The JsonSerializerOptions and JsonTypeInfoResolver get lazy initialized in JsonOptions here:

https://github.com/dotnet/aspnetcore/blob/9427854d4f10f62a5f7f89988ea367b278ea1703/src/Http/Http.Extensions/src/JsonOptions.cs#L27-L34

when `IsReflectionEnabledByDeafult=false`, the Resolver may be null. We should fix the thread safety concerns all up (RDG, RDF, and MVC) with https://github.com/dotnet/aspnetcore/issues/49849.

Reply:
It would depend on how `DefaultSerializerOptions` is defined. If it contains a resolver it would get picked up by the copy constructor.

Apropos, does the singleton need to be `JsonOptions` here. Seems it might as well be a `JsonSerializerOptions`.

Reply:
> It would depend on how DefaultSerializerOptions is defined. If it contains a resolver it would get picked up by the copy constructor.

`DefaultSerializerOptions` doesn't initialize a resolver AFAICT (it's just `new JsonSerlizerOptions(JsonDefaults.Web`) so I think being explicit here helps.

> Apropos, does the singleton need to be JsonOptions here. Seems it might as well be a JsonSerializerOptions.

True.

Reply:
> It would depend on how DefaultSerializerOptions is defined.

It is part of the above linked code. I missed a couple lines.

https://github.com/dotnet/aspnetcore/blob/9427854d4f10f62a5f7f89988ea367b278ea1703/src/Http/Http.Extensions/src/JsonOptions.cs#L16-L34

> Apropos, does the singleton need to be JsonOptions here. Seems it might as well be a JsonSerializerOptions.

It should be using the `DefaultSerializerOptions` that is in JsonOptions because this is where our default JsonSerializerOptions is defined. We shouldn't be defining a new one here in generated code.

Reply:
Couldn't it just be referencing JsonOptions.DefaultSerializerOptions directly without wrapping it in a JsonOptions then?